src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -471,15 +480,76 @@ public ManagedObjectWrapperReleaser(ManagedObjectWrapper* wrapper)
         internal unsafe class NativeObjectWrapper
         {
             internal IntPtr _externalComObject;
+            private IntPtr _inner;
             private ComWrappers _comWrappers;
             internal GCHandle _proxyHandle;
+            internal GCHandle _proxyHandleTrackingResurrection;
+            internal GCHandle _nativeObjectWrapperWeakHandle;
+
+            private readonly bool _aggregatedManagedObjectWrapper;
+            internal readonly bool _fromTrackerRuntime;
+            private IntPtr _trackerObject;
+            private readonly bool _releaseTrackerObject;
+            private int _trackerObjectDisconnected;
+
+            internal readonly IntPtr _contextToken;
 
-            public NativeObjectWrapper(IntPtr externalComObject, ComWrappers comWrappers, object comProxy)
+            public IntPtr TrackerObject => (_trackerObject == IntPtr.Zero || _trackerObjectDisconnected == 1) ? IntPtr.Zero : _trackerObject;
+
+            public NativeObjectWrapper(IntPtr externalComObject, IntPtr inner, ComWrappers comWrappers, object comProxy, CreateObjectFlags flags)
             {
                 _externalComObject = externalComObject;
+                _inner = inner;
                 _comWrappers = comWrappers;
-                Marshal.AddRef(externalComObject);
                 _proxyHandle = GCHandle.Alloc(comProxy, GCHandleType.Weak);
+
+                // We have a separate handle tracking resurrection as we want to make sure
+                // we clean up the NativeObjectWrapper only after the RCW has been finalized
+                // due to it can access the native object in the finalizer. At the same time,
+                // we want other callers which are using _proxyHandle such as the rcw cache to
+                // see the object as not alive once it is eligible for finalization.
+                _proxyHandleTrackingResurrection = GCHandle.Alloc(comProxy, GCHandleType.WeakTrackResurrection);
+
+                _nativeObjectWrapperWeakHandle = GCHandle.Alloc(this, GCHandleType.Weak);
+
+                if (flags.HasFlag(CreateObjectFlags.TrackerObject))
+                {
+                    if (Marshal.QueryInterface(externalComObject, in IID_IReferenceTracker, out IntPtr trackerObject) == 0 &&
+                        trackerObject != IntPtr.Zero)
```

Comment:
```suggestion
                        trackerObject != IntPtr.Zero)
```

I would remove this. If the QI returns `S_OK` and `null`, I want this to fail hard.

Reply:
Feel free to assert, but this is breaking a well established contract and is simply unnecessary overhead.

Reply:
Removed check.